### PR TITLE
Add info about TF_DETERMINISTIC_OPS to version 2.1 release notes

### PR DESCRIPTION
**Pull request by duncanriach**
**This pull request is specifically targeted at the r2.1 branch.**

This adds information about the environment variable `TF_DETERMINISTIC_OPS`, which is first introduced in version 2.1, to the associated release notes.

For more information, see [PR 31465](https://github.com/tensorflow/tensorflow/pull/31465) that implemented this functionality.

### CODE REVIEW

**Review by geetachavan1 — DISMISSED** [review]

**Comment by aaroey** [comment]
I'm not familiar with this, @chsigg @sanjoy could you help to take a look?

File: RELEASE.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ TensorFlow 2.1 will be the last TF release supporting Python 2. Python 2 support
   * Changes rebatching for `tf.data datasets` + distribution strategies for better performance.   Note that the dataset also behaves slightly differently, in that the rebatched dataset cardinality will always be a multiple of the number of replicas.
 * `TensorRT`
   * [TensorRT 6.0](https://developer.nvidia.com/tensorrt#tensorrt-whats-new) is now supported and enabled by default. This adds support for more TensorFlow ops including Conv3D, Conv3DBackpropInputV2, AvgPool3D, MaxPool3D, ResizeBilinear, and ResizeNearestNeighbor. In addition, the TensorFlow-TensorRT python conversion API is exported as `tf.experimental.tensorrt.Converter`.
+  * Environment variable `TF_DETERMINISTIC_OPS` added. When set to "true" or "1", this environment variable makes `tf.nn.bias_add` operate deterministically (i.e. reproducibly) when XLA JIT compilation is *not* enabled. It also makes cuDNN convolution and max-pooling operate deterministically. This makes Keras Conv*D and MaxPool*D layers operate deterministically in both the forward and backward directions when running on a CUDA-enabled GPU.
```
Comment (sanjoy) [inline RELEASE.md]: @timshen91 shouldn't this also make XLAs' convolution deterministic?  And XLA's max-pool is already deterministic I think.

Comment (duncanriach) [reply RELEASE.md]: @sanjoy, `TF_DETERMINISTIC_OPS` does make XLA JIT compiled cuDNN convolution and max-pooling deterministic, just as `TF_CUDNN_DETERMINISTIC` does. However, the current (temporary) solution to make `tf.nn.bias_add` deterministic under `TF_DETERMINISTIC_OPS` is undone by XLA JIT compilation. When `tf.nn.bias_add` is made deterministic at the CUDA kernel level (which I plan to do), then XLA JIT compilation will not be able to undo the deterministic mode.

For v2.1, this functionality is finalized. If someone can suggest to me a way to locally prevent JIT optimization of only the Python code that provides a deterministic implementation of `tf.nn.bias_add` then that could be added for v2.2, and all determinism promised by `TF_DETERMINISTIC_OPS` could be delivered with XLA JIT compilation enabled.

Comment (sanjoy) [reply RELEASE.md]: > However, the current (temporary) [solution](https://github.com/tensorflow/tensorflow/pull/31465) to make [`tf.nn.bias_add`](https://github.com/tensorflow/tensorflow/blob/v2.1.0-rc0/tensorflow/python/ops/nn_ops.py#L2699) deterministic under `TF_DETERMINISTIC_OPS` is [undone by XLA JIT compilation](https://github.com/tensorflow/tensorflow/pull/31465#discussion_r333793547).

I'm investigating why this happens but given what you said the notes LGTM.

Comment (duncanriach) [reply RELEASE.md]: I don't know if this helps, and I'm not sure if I've written this anywhere else:

At the highest level, my working hypothesis was that the compiler was able to (1) find the original, non-deterministic op code, (2) recognize that it was arithmetically equivalent to the alternative grouping of deterministic ops, and (3) switch in the original op because it has higher performance.

However, I didn't the use profiler to confirm that the old op's CUDA kernel was re-appearing with XLA JIT compilation enabled, and also I understand very little about the mechanics of the XLA JIT compiler and how it would have been able to accomplish this.

Comment (sanjoy) [reply RELEASE.md]: The HLO (XLA IR) we get out of TF has a `reduce` operation in it and reductions on XLA GPU are not deterministic: https://gist.github.com/sanjoy/bc9636188eb59bf42b211aa1e4614068

@cheshire is working on a fix for this.  Can you share some context on why you need this?  That will help us prioritize implementing deterministic reductions.

Comment (duncanriach) [reply RELEASE.md]: @sanjoy: Awesome! That makes so much more sense to me. Great to know. Thank you.

Regarding, "Can you share some context on why you need this?" Is this a question for me? If so, are you asking why deterministic `tf.nn.bias_add` is important in TensorFlow in general?

Comment (sanjoy) [reply RELEASE.md]: > Is this a question for me? If so, are you asking why deterministic `tf.nn.bias_add` is important in TensorFlow in general?

Yes.  I realize that there are general reasons why determinism is important (debuggability comes to mind) but if you have a more specific use case we'd like to know about it.

**Review by sanjoy — APPROVED** [review]
lgtm

File: RELEASE.md (hunk 1 above)
Comment (duncanriach) [reply RELEASE.md]: There are four main areas I think about related to determinism:

Determinism is important for [1] experimentation (including hyper-parameter tuning), [2] debugging (as you noted), and [3] being able to detect and prevent regressions.

On the last point, for example, I discovered a serious bug in the simple LSTM cell in tensorflow/models while visually inspecting the code. Despite this bug, in the small configuration used for regression-testing the model happened to be able to train to within the expected range of accuracy. So the bug didn't fail the test and was not easily discovered.

However, the biggest current factor driving the need for determinism is [4] deep learning solutions being deployed into production in safety-critical applications and markets (which are often government-regulated). This applies to any robotic application with proximity to humans, such as autonomous vehicles, and also to healthcare applications, such as medical diagnosis and imaging.

I don't know the full history here, but non-deterministic `reduce_sum` and `reduce_mean` operators in TensorFlow were historically problematic. There were a lot of complaints about them in places like TF issues and on Stack Overflow. Non-determinism from `reduce_sum` and `reduce_mean` also often resulted in non-determinism in back-propagation because reduction is the way to back-prop gradients through an implicit broadcast. However, since at least v1.12, `reduce_sum` and `reduce_mean` have been operating deterministically.

Coincidentally, the (temporary) solution (referred to by this current pull request) to provide a deterministic mode for `tf.nn.bias_add` relies on deterministic functionality of the reduction that is used to back-propagate the gradients through the implicit broadcast that it uses.

It's a relief to know what the source of non-determinism when using `tf.nn.bias_add` with both `TF_DETERMINISTIC_OPS` and XLA enabled is, and to know that @cheshire is working on a fix for it. Please will you keep me informed about the status of this so that I can reflect it in the public documentation I maintain [here](https://github.com/NVIDIA/tensorflow-determinism)?

Also, `xla_enable_strict_auto_jit` should be set to `True` for this test. for [this test](https://github.com/tensorflow/tensorflow/blob/v2.1.0-rc0/tensorflow/python/kernel_tests/BUILD#L1680).

Comment (cheshire) [reply RELEASE.md]: @duncandean I have landed deterministic reductions in XLA behind the flag (XLA_FLAGS=--xla_gpu_deterministic_reductions), this commit will appear in the repository shortly. We are currently evaluating the performance to see whether we can switch it on by default.

Comment (duncanriach) [reply RELEASE.md]: @cheshire,

Wow! Thank you so much for doing that and also for letting me know.

Yes, please let me know if that becomes the default.

PS. You tagged the wrong Duncan.

Comment (duncanriach) [reply RELEASE.md]: Also, please let me know if it gets cherry-picked into version 2.1 or if it will end up in version 2.2.

Comment (cheshire) [reply RELEASE.md]: I don't think we'll cherry-pick it to 2.1. I expect it to end up in 2.2. If we are lucky, it might become the default by then, but no promises.

Comment (duncanriach) [reply RELEASE.md]: Sounds good. I have added [info about this solution](https://github.com/NVIDIA/tensorflow-determinism#confirmed-current-gpu-specific-sources-of-non-determinism-with-solutions) to the tensorflow-determinism README. I also added your name to the [credits](https://github.com/NVIDIA/tensorflow-determinism#credits).

Comment (cheshire) [reply RELEASE.md]: @duncanriach As of https://github.com/tensorflow/tensorflow/commit/8b7a3db0b6e09415b5640be4986fb4d7c6e5209a reduction emitter respects TF_DETERMINISTIC_OPS. On our benchmark, it's performance is within 0.9x of the emitter which uses atomics.

Comment (duncanriach) [reply RELEASE.md]: Awesome! Thanks for doing this and for letting me know, @cheshire. I plan to do the following:
  * Update (tensorflow-determinism)[https://github.com/NVIDIA/tensorflow-determinism] to reflect this feature enhancement (coming in TF 2.2).
  * Add info about this to the TF 2.2 release notes when that snaps.
  * Submit a PR (hopefully before the 2.2 snap) to (1) make `TF_DETERMINISTIC_OPS` cached and sticky, to exactly match its functionality elsewhere and (2) enable the deterministic `bias_add` (kernel test)[https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/kernel_tests/BUILD#L1680] for XLA.

Comment (cheshire) [reply RELEASE.md]: Cool. Once the performance is closer to the atomic version, I hope we can even switch to deterministic reductions by default.

Comment (duncanriach) [reply RELEASE.md]: Nice. Thank so much, @cheshire!
